src/content/developers/docs/consensus-mechanisms/pow/mining-algorithms/dagger-hashamoto/index.md

```diff
@@ -0,0 +1,327 @@
+
+```python
+def quick_hashimoto(seed, dagsize, params, header, nonce):
+    m = dagsize // 2
+    mix = sha3(nonce + header)
+    for _ in range(params["accesses"]):
+        mix ^= quick_calc(params, seed, m + (mix % 2**64) % m)
+    return dbl_sha3(mix)
+```
+
+## Mining and Verifying
+
+Now, let us put it all together into the mining algo:
+
+```python
+def mine(daggerset, params, block):
+    from random import randint
+    nonce = randint(0, 2**64)
+    while 1:
+        result = hashimoto(daggerset, get_dagsize(params, block),
+                           params, decode_int(block.prevhash), nonce)
+        if result * params["diff"] < 2**256:
+            break
+        nonce += 1
+        if nonce >= 2**64:
+            nonce = 0
+    return nonce
+```
+
+Here is the verification algorithm:
+
+```python
+def verify(daggerset, params, block, nonce):
+    result = hashimoto(daggerset, get_dagsize(params, block),
+                       params, decode_int(block.prevhash), nonce)
+    return result * params["diff"] < 2**256
+```
+
+Light-client friendly verification:
+
+```python
+def light_verify(params, header, nonce):
+    seedset = get_seedset(params, block)
+    result = quick_hashimoto(seedset["front_hash"], get_dagsize(params, block),
+                             params, decode_int(block.prevhash), nonce)
+    return result * params["diff"] < 2**256
+```
+
+Also, note that Dagger Hashimoto imposes additional requirements on the block header:
+
+- For two-layer verification to work, a block header must have both the nonce and the middle value pre-sha3
+- Somewhere, a block header must store the sha3 of the current seedset
+
+## Appendix
+
+As noted above, the RNG used for DAG generation relies on some results from number theory. First, we provide assurance that the Lehmer RNG that is the basis for the `picker` variable has a wide period. Second, we show that `pow(x,3,P)` will not map `x` to `1` or `P-1` provided `x ∈ [2,P-2]` to start. Finally, we show that `pow(x,3,P)` has a low collision rate when treated as a hashing function.
+
+### Lehmer Random Number Generator
+
+While the `produce_dag` function does not need to produce unbiased random numbers, a potential threat is that `seed**i % P` only takes on a handful of values. This could provide an advantage to miners recognizing the pattern over those that do not.
+
+To avoid this, a result from number theory is appealed to. A [_Safe Prime_](https://en.wikipedia.org/wiki/Safe_prime) is defined to be a prime `P` such that `(P-1)/2` is also prime. The _order_ of a member `x` of the [multiplicative group](https://en.wikipedia.org/wiki/Multiplicative_group_of_integers_modulo_n) `ℤ/nℤ` is defined to be the minimal `m` such that <center><pre>xᵐ mod P ≡ 1</pre></center>
```

Comment:
Is there a need for centering here?
```suggestion
To avoid this, a result from number theory is appealed to. A [_Safe Prime_](https://en.wikipedia.org/wiki/Safe_prime) is defined to be a prime `P` such that `(P-1)/2` is also prime. The _order_ of a member `x` of the [multiplicative group](https://en.wikipedia.org/wiki/Multiplicative_group_of_integers_modulo_n) `ℤ/nℤ` is defined to be the minimal `m` such that <pre>xᵐ mod P ≡ 1</pre>
```